<a href="https://colab.research.google.com/github/amandarice97/data-analytics/blob/main/danl-m1-hw4-Rice-Amanda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Load Libraries
Import the pandas library.

In [1]:
import pandas as pd # import the pandas libraray

# Part 1
Below is nyc_payroll DataFrame that reads the file nyc_payroll.csv containing data of how the New York City’s budget is being spent on salary and overtime pay for all municipal employees (Source: NYC OpenData).

In [3]:
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv')
nyc_payroll

Fiscal Year  Payroll Number                     Agency Name Last Name  \
0            2022             740   DEPARTMENT OF EDUCATION ADMIN     DAVIS   
1            2022             746    DEPT OF ED PER DIEM TEACHERS      ZHOU   
2            2022             747  DEPT OF ED PER SESSION TEACHER     LOPEZ   
3            2022             742          DEPT OF ED PEDAGOGICAL   KENNEDY   
4            2022             742          DEPT OF ED PEDAGOGICAL   LICHVAR   
...           ...             ...                             ...       ...   
1995         2023             300  BOARD OF ELECTION POLL WORKERS    GARNER   
1996         2023             742          DEPT OF ED PEDAGOGICAL  FIGUEROA   
1997         2023             747  DEPT OF ED PER SESSION TEACHER   DIFORTE   
1998         2023             747  DEPT OF ED PER SESSION TEACHER    BIONDO   
1999         2023             827        DEPARTMENT OF SANITATION     ARIAS   

     First Name Mid Init Agency Start Date Work Location Borough  \
0         OSCAR      NaN        07/06/2021                 BRONX   
1           JUN      NaN        07/01/2020             MANHATTAN   
2       YERIBEL        E        12/11/2001             MANHATTAN   
3          VERA        R        04/08/2016             MANHATTAN   
4        AMANDA        C        02/01/2007             MANHATTAN   
...         ...      ...               ...                   ...   
1995   GEOFFREY        O        01/01/2020             MANHATTAN   
1996    JESSICA        A        02/08/2006             MANHATTAN   
1997   DANIELLE        N        09/05/2017             MANHATTAN   
1998   STEFANIE      NaN        10/03/2011             MANHATTAN   
1999    ELPIDIO      NaN        07/11/2022             MANHATTAN   

                                      Title Description  \
0     SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
1                                    TEACHER-GENERAL ED   
2                                  TEACHER- PER SESSION   
3                                               TEACHER   
4                             TEACHER SPECIAL EDUCATION   
...                                                 ...   
1995                                    ELECTION WORKER   
1996                                 TEACHER ASSIGNED A   
1997                               TEACHER- PER SESSION   
1998                               TEACHER- PER SESSION   
1999                                  SANITATION WORKER   

     Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                        CEASED     53820.00  per Annum          805.0   
1                        ACTIVE       110.29    per Day            0.0   
2                        ACTIVE        33.18    per Day            0.0   
3                        ACTIVE     79687.00  per Annum            0.0   
4                        ACTIVE    109852.00  per Annum            0.0   
...                         ...          ...        ...            ...   
1995                     ACTIVE         1.00   per Hour            0.0   
1996                     ACTIVE    109852.00  per Annum            0.0   
1997                     ACTIVE        33.18    per Day            0.0   
1998                     ACTIVE        33.18    per Day            0.0   
1999                     ACTIVE     44064.00  per Annum         2000.0   

      Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0               24469.41       0.0           0.00             0.00  
1                3387.59       0.0           0.00             0.00  
2               20242.24       0.0           0.00           474.32  
3               74904.27       0.0           0.00         -4474.44  
4              109852.08       0.0           0.00          5380.44  
...                  ...       ...            ...              ...  
1995              928.00       0.0           0.00             0.00  
1996           109852.08       0.0           0.00             0.00  
1997             1824

# Variable Description
- Fiscal Year: Fiscal Year;
- Payroll Number: Payroll Number;
- Agency Name: The Payroll agency that the employee works for;
- Last Name: Last name of employee;
- First Name: First name of employee;
- Mid Init: Middle initial of employee;
- Agency Start Date: Date which employee began working for their current agency;
- Work Location Borough: Borough of employee’s primary work location;
- Title Description: Civil service title description of the employee;
- Leave Status as of June 30: Status of employee as of the close of the relevant fiscal year;
- Base Salary: Base Salary assigned to the employee;
- Pay Basis: Lists whether the employee is paid on an hourly, per diem or annual basis;
- Regular Hours: Number of regular hours employee worked in the fiscal year;
- Regular Gross Paid: The amount paid to the employee for base salary during the fiscal year;
- OT Hours: Overtime Hours worked by employee in the fiscal year;
- Total OT Paid: Total overtime pay paid to the employee in the fiscal year;
- Total Other Pay: Includes any compensation in addition to gross salary and overtime pay, ie Differentials, lump sums, uniform allowance, meal allowance, retroactive pay increases, settlement amounts, and bonus pay, if applicable.

# Question 1
How many employees have a “Base Salary” within the top 10% of the DataFrame?

Hint: The .quantile() method can be useful.

*Answer*

In [20]:
nyc_payroll['Base Salary'].quantile(0.10) # use .quantile(0.10) to identify employees who have a 'Base Salary' within the top 10% of the DataFrame

16.88

# Question 2
Filter the DataFrame for employees who have “OT Hours” greater than 0 but less than 100, and their “Leave Status as of June 30” is “ACTIVE”.

*Answer*

In [21]:
nyc_payroll["Leave Status as of June 30"] == "ACTIVE" # filter 'Leave Status as of June 30' is 'ACTIVE'

0       True
1       True
2       True
3       True
4       True
        ... 
1995    True
1996    True
1997    True
1998    True
1999    True
Name: Leave Status as of June 30, Length: 2000, dtype: bool

In [22]:
nyc_payroll[ nyc_payroll["Leave Status as of June 30"] == "ACTIVE"] # display filter results

Fiscal Year  Payroll Number                     Agency Name Last Name  \
0            2022             740   DEPARTMENT OF EDUCATION ADMIN     DAVIS   
1            2022             746    DEPT OF ED PER DIEM TEACHERS      ZHOU   
2            2022             747  DEPT OF ED PER SESSION TEACHER     LOPEZ   
3            2022             742          DEPT OF ED PEDAGOGICAL   KENNEDY   
4            2022             742          DEPT OF ED PEDAGOGICAL   LICHVAR   
...           ...             ...                             ...       ...   
1995         2023             300  BOARD OF ELECTION POLL WORKERS    GARNER   
1996         2023             742          DEPT OF ED PEDAGOGICAL  FIGUEROA   
1997         2023             747  DEPT OF ED PER SESSION TEACHER   DIFORTE   
1998         2023             747  DEPT OF ED PER SESSION TEACHER    BIONDO   
1999         2023             827        DEPARTMENT OF SANITATION     ARIAS   

     First Name Mid Init Agency Start Date Work Location Borough  \
0         OSCAR      NaN        07/06/2021                 BRONX   
1           JUN      NaN        07/01/2020             MANHATTAN   
2       YERIBEL        E        12/11/2001             MANHATTAN   
3          VERA        R        04/08/2016             MANHATTAN   
4        AMANDA        C        02/01/2007             MANHATTAN   
...         ...      ...               ...                   ...   
1995   GEOFFREY        O        01/01/2020             MANHATTAN   
1996    JESSICA        A        02/08/2006             MANHATTAN   
1997   DANIELLE        N        09/05/2017             MANHATTAN   
1998   STEFANIE      NaN        10/03/2011             MANHATTAN   
1999    ELPIDIO      NaN        07/11/2022             MANHATTAN   

                                      Title Description  \
0     SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
1                                    TEACHER-GENERAL ED   
2                                  TEACHER- PER SESSION   
3                                               TEACHER   
4                             TEACHER SPECIAL EDUCATION   
...                                                 ...   
1995                                    ELECTION WORKER   
1996                                 TEACHER ASSIGNED A   
1997                               TEACHER- PER SESSION   
1998                               TEACHER- PER SESSION   
1999                                  SANITATION WORKER   

     Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                        ACTIVE     53820.00  per Annum          805.0   
1                        ACTIVE       110.29    per Day            0.0   
2                        ACTIVE        33.18    per Day            0.0   
3                        ACTIVE     79687.00  per Annum            0.0   
4                        ACTIVE    109852.00  per Annum            0.0   
...                         ...          ...        ...            ...   
1995                     ACTIVE         1.00   per Hour            0.0   
1996                     ACTIVE    109852.00  per Annum            0.0   
1997                     ACTIVE        33.18    per Day            0.0   
1998                     ACTIVE        33.18    per Day            0.0   
1999                     ACTIVE     44064.00  per Annum         2000.0   

      Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0               24469.41       0.0           0.00             0.00  
1                3387.59       0.0           0.00             0.00  
2               20242.24       0.0           0.00           474.32  
3               74904.27       0.0           0.00         -4474.44  
4              109852.08       0.0           0.00          5380.44  
...                  ...       ...            ...              ...  
1995              928.00       0.0           0.00             0.00  
1996           109852.08       0.0           0.00             0.00  
1997             1824

In [23]:
greater_than_0_hrs = nyc_payroll["OT Hours"] > 0
less_than_100_hrs = nyc_payroll["OT Hours"] < 100
nyc_payroll[ greater_than_0_hrs & less_than_100_hrs] # '&' is 'and' operator # filter by a condition

Fiscal Year  Payroll Number                   Agency Name  Last Name  \
19           2022             841  DEPARTMENT OF TRANSPORTATION     MURRAY   
34           2022             996         NYC HOUSING AUTHORITY     GIRAUD   
48           2022             846    DEPT OF PARKS & RECREATION     ORMSBY   
75           2022             841  DEPARTMENT OF TRANSPORTATION  RODRIGUEZ   
81           2022              69   HRA/DEPT OF SOCIAL SERVICES      JAMES   
...           ...             ...                           ...        ...   
1962         2023             846    DEPT OF PARKS & RECREATION      BANKS   
1970         2023              56             POLICE DEPARTMENT     COTHIA   
1974         2023             846    DEPT OF PARKS & RECREATION      BLACK   
1977         2023             846    DEPT OF PARKS & RECREATION     GARDEN   
1986         2023              69   HRA/DEPT OF SOCIAL SERVICES     MILLER   

     First Name Mid Init Agency Start Date Work Location Borough  \
19        LEROY      NaN        06/02/2008              RICHMOND   
34        WAYNE        M        08/30/2010                QUEENS   
48       CLAUDE      NaN        06/27/1989              BROOKLYN   
75        HENRY      NaN        11/23/2015             MANHATTAN   
81         DALE      NaN        05/29/2007             MANHATTAN   
...         ...      ...               ...                   ...   
1962     ANDREW      NaN        07/08/2022             MANHATTAN   
1970    NADEIGE      NaN        07/08/2008                 BRONX   
1974     ISAIAH      NaN        04/27/2023                QUEENS   
1977     DIEOSA        L        09/16/2022             MANHATTAN   
1986      KAREN      NaN        12/12/1994              BROOKLYN   

                              Title Description Leave Status as of June 30  \
19                                 MARINE OILER                     ACTIVE   
34                                    INSPECTOR                     ACTIVE   
48    ADMINISTRATIVE PARKS & RECREATION MANAGER                     ACTIVE   
75                        COMMUNITY COORDINATOR                     ACTIVE   
81                                   CASEWORKER                     ACTIVE   
...                                         ...                        ...   
1962                   JOB TRAINING PARTICIPANT                     ACTIVE   
1970                             POLICE OFFICER                     ACTIVE   
1974                   JOB TRAINING PARTICIPANT                     ACTIVE   
1977                   JOB TRAINING PARTICIPANT                     ACTIVE   
1986                     ELIGIBILITY SPECIALIST                     ACTIVE   

      Base Salary  Pay Basis  Regular Hours  Regular Gross Paid  OT Hours  \
19       61555.00  per Annum           0.00                0.00      8.00   
34       77091.00  per Annum        2160.00            79836.82      8.00   
48       96444.00  per Annum        1820.00            96179.72     91.00   
75       62215.00  per Annum        1820.00            62044.58     49.75   
81       55614.00  per Annum        1820.00            55461.64     48.00   
...           ...        ...            ...                 ...       ...   
1962        16.39   per Hour        1332.93            21846.79     71.25   
1970    101590.00  per Annum        2080.00            85058.22      6.65   
1974        16.88   per Hour         231.50             3625.72      9.50   
1977        16.88   per Hour        1352.33            22196.77     14.00   
1986     48293.00  per Annum        1820.00            44568.59     83.00   

      Total OT Paid  Total Other Pay  
19           433.04             0.00  
34           455.86          2308.50  
48          5765.34          3030.62  
75          1767.71          1427.45  
81          1989.66          8757.05  
...             ...              ...  
1962        1512.00          1744.11  
1970         452.34         14823.29  
1974         220.16            49.14  


# Question 3
Find the unique job titles in the “DEPARTMENT OF EDUCATION ADMIN” agency and count how many there are.

*Answer*

**There are 26 unique job titles for Dept. of Edu. Admin. employees. **

In [27]:
Dept_of_Edu_Admin = (
    nyc_payroll["Agency Name"] == "DEPARTMENT OF EDUCATION ADMIN"
)
Dept_of_Edu_Admin

0        True
1       False
2       False
3       False
4       False
        ...  
1995    False
1996    False
1997    False
1998    False
1999    False
Name: Agency Name, Length: 2000, dtype: bool

In [28]:
nyc_payroll[ Dept_of_Edu_Admin ]['Agency Name'].value_counts() # num. of employees working for the Dept. of Edu. Admin.

DEPARTMENT OF EDUCATION ADMIN    73
Name: Agency Name, dtype: int64

In [31]:
nyc_payroll[ Dept_of_Edu_Admin ]['Title Description'].nunique() # num. of unique job titles (26) for employees working for the Dep. of Edu. Admin.

26

# Question 4
Identify the employee(s) with the highest “Total OT Paid” in the DataFrame.
Include their “First Name”, “Last Name”, and “Total OT Paid”.

*Answer*

**See below for answer.**

In [32]:
nyc_payroll

Fiscal Year  Payroll Number                     Agency Name Last Name  \
0            2022             740   DEPARTMENT OF EDUCATION ADMIN     DAVIS   
1            2022             746    DEPT OF ED PER DIEM TEACHERS      ZHOU   
2            2022             747  DEPT OF ED PER SESSION TEACHER     LOPEZ   
3            2022             742          DEPT OF ED PEDAGOGICAL   KENNEDY   
4            2022             742          DEPT OF ED PEDAGOGICAL   LICHVAR   
...           ...             ...                             ...       ...   
1995         2023             300  BOARD OF ELECTION POLL WORKERS    GARNER   
1996         2023             742          DEPT OF ED PEDAGOGICAL  FIGUEROA   
1997         2023             747  DEPT OF ED PER SESSION TEACHER   DIFORTE   
1998         2023             747  DEPT OF ED PER SESSION TEACHER    BIONDO   
1999         2023             827        DEPARTMENT OF SANITATION     ARIAS   

     First Name Mid Init Agency Start Date Work Location Borough  \
0         OSCAR      NaN        07/06/2021                 BRONX   
1           JUN      NaN        07/01/2020             MANHATTAN   
2       YERIBEL        E        12/11/2001             MANHATTAN   
3          VERA        R        04/08/2016             MANHATTAN   
4        AMANDA        C        02/01/2007             MANHATTAN   
...         ...      ...               ...                   ...   
1995   GEOFFREY        O        01/01/2020             MANHATTAN   
1996    JESSICA        A        02/08/2006             MANHATTAN   
1997   DANIELLE        N        09/05/2017             MANHATTAN   
1998   STEFANIE      NaN        10/03/2011             MANHATTAN   
1999    ELPIDIO      NaN        07/11/2022             MANHATTAN   

                                      Title Description  \
0     SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
1                                    TEACHER-GENERAL ED   
2                                  TEACHER- PER SESSION   
3                                               TEACHER   
4                             TEACHER SPECIAL EDUCATION   
...                                                 ...   
1995                                    ELECTION WORKER   
1996                                 TEACHER ASSIGNED A   
1997                               TEACHER- PER SESSION   
1998                               TEACHER- PER SESSION   
1999                                  SANITATION WORKER   

     Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                        ACTIVE     53820.00  per Annum          805.0   
1                        ACTIVE       110.29    per Day            0.0   
2                        ACTIVE        33.18    per Day            0.0   
3                        ACTIVE     79687.00  per Annum            0.0   
4                        ACTIVE    109852.00  per Annum            0.0   
...                         ...          ...        ...            ...   
1995                     ACTIVE         1.00   per Hour            0.0   
1996                     ACTIVE    109852.00  per Annum            0.0   
1997                     ACTIVE        33.18    per Day            0.0   
1998                     ACTIVE        33.18    per Day            0.0   
1999                     ACTIVE     44064.00  per Annum         2000.0   

      Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0               24469.41       0.0           0.00             0.00  
1                3387.59       0.0           0.00             0.00  
2               20242.24       0.0           0.00           474.32  
3               74904.27       0.0           0.00         -4474.44  
4              109852.08       0.0           0.00          5380.44  
...                  ...       ...            ...              ...  
1995              928.00       0.0           0.00             0.00  
1996           109852.08       0.0           0.00             0.00  
1997             1824

In [42]:
nyc_payroll.columns # identify column headings in nyc_payroll

Index(['Fiscal Year', 'Payroll Number', 'Agency Name', 'Last Name',
       'First Name', 'Mid Init', 'Agency Start Date', 'Work Location Borough',
       'Title Description', 'Leave Status as of June 30', 'Base Salary',
       'Pay Basis', 'Regular Hours', 'Regular Gross Paid', 'OT Hours',
       'Total OT Paid', 'Total Other Pay'],
      dtype='object')

In [43]:
nyc_payroll = nyc_payroll[ ['First Name', 'Last Name', 'Total OT Paid'] ] # remove unnecessary columns
nyc_payroll

First Name Last Name  Total OT Paid
0         OSCAR     DAVIS           0.00
1           JUN      ZHOU           0.00
2       YERIBEL     LOPEZ           0.00
3          VERA   KENNEDY           0.00
4        AMANDA   LICHVAR           0.00
...         ...       ...            ...
1995   GEOFFREY    GARNER           0.00
1996    JESSICA  FIGUEROA           0.00
1997   DANIELLE   DIFORTE           0.00
1998   STEFANIE    BIONDO           0.00
1999    ELPIDIO     ARIAS        5659.77

[2000 rows x 3 columns]

In [44]:
nyc_payroll['Total OT Paid'].nlargest(10) # identify 10 highest amounts for 'Total OT Paid' using .nlargest(10)

1199    126184.02
1294     92892.79
462      76678.61
932      75099.43
1840     75091.43
1680     74059.44
666      72664.59
162      70651.05
1172     68725.63
603      65753.36
Name: Total OT Paid, dtype: float64

In [45]:
high_total_OT_paid = nyc_payroll['Total OT Paid'] > 65753 # filter by a condition
nyc_payroll[ high_total_OT_paid ]

First Name       Last Name  Total OT Paid
162     JEFFREY       CALLAGHAN       70651.05
462      GLORIA          CHAVEZ       76678.61
603     GREGORY         GENTILE       65753.36
666        JOHN        SYNOWIEC       72664.59
932         NaN             NaN       75099.43
1172    DOMINIC        ROBINSON       68725.63
1199   EMMANUEL  KOUTSOULIDAKIS      126184.02
1294     ROBERT         RODGERS       92892.79
1680        NaN             NaN       74059.44
1840     REGINA           ADAMI       75091.43

# Part 2
Below is spotify DataFrame that reads the file spotify_all.csv containing data of Spotify users’ playlist information (Source: Spotify Million Playlist Dataset Challenge).

In [46]:
import pandas as pd # import pandas as pd

In [118]:
spotify = pd.read_csv('https://bcdanl.github.io/data/spotify_all.csv')
spotify

pid playlist_name  pos        artist_name  \
0            0    Throwbacks    0      Missy Elliott   
1            0    Throwbacks    1     Britney Spears   
2            0    Throwbacks    2            Beyoncé   
3            0    Throwbacks    3  Justin Timberlake   
4            0    Throwbacks    4             Shaggy   
...        ...           ...  ...                ...   
198000  999998            ✝️    6       Chris Tomlin   
198001  999998            ✝️    7       Chris Tomlin   
198002  999998            ✝️    8            Crowder   
198003  999998            ✝️    9       Chris Tomlin   
198004  999998            ✝️   10   Caleb and Kelsey   

                                        track_name  duration_ms  \
0       Lose Control (feat. Ciara & Fat Man Scoop)       226863   
1                                            Toxic       198800   
2                                    Crazy In Love       235933   
3                                   Rock Your Body       267266   
4                                     It Wasn't Me       227600   
...                                            ...          ...   
198000                                   Waterfall       209573   
198001                                    The Roar       220106   
198002        Lift Your Head Weary Sinner (Chains)       224666   
198003                                We Fall Down       280960   
198004      10,000 Reasons / What a Beautiful Name       178189   

                                            album_name  
0                                         The Cookbook  
1                                          In The Zone  
2         Dangerously In Love (Alben für die Ewigkeit)  
3                                            Justified  
4                                             Hot Shot  
...                                                ...  
198000                                    Love Ran Red  
198001                                    Love Ran Red  
198002                                    Neon Steeple  
198003  How Great Is Our God: The Essential Collection  
198004          10,000 Reasons / What a Beautiful Name  

[198005 rows x 7 columns]

# Variable Description
- pid: playlist ID; unique ID for playlist
- playlist_name: a name of playlist
- pos: a position of the track within a playlist (starting from 0)
- artist_name: name of the track’s primary artist
- track_name: name of the track
- duration_ms: duration of the track in milliseconds
- album_name: name of the track’s album

# Question 5
Find the five most popular songs.
- A value of a song is defined as a combination of a artist_name value and a track_name value.
- Who are artists for those five most popular song?

*Answer*

**Five Most Popular Songs & thier Artist:**
- Drake (One Dance)
- Kendrick Lamar (HUMBLE)
- The Chainsmokers (Closer)
- DRAM (Broccoli (feat. Lil Yachty))
- Post Malone (Congratulations)


In [119]:
spotify

pid playlist_name  pos        artist_name  \
0            0    Throwbacks    0      Missy Elliott   
1            0    Throwbacks    1     Britney Spears   
2            0    Throwbacks    2            Beyoncé   
3            0    Throwbacks    3  Justin Timberlake   
4            0    Throwbacks    4             Shaggy   
...        ...           ...  ...                ...   
198000  999998            ✝️    6       Chris Tomlin   
198001  999998            ✝️    7       Chris Tomlin   
198002  999998            ✝️    8            Crowder   
198003  999998            ✝️    9       Chris Tomlin   
198004  999998            ✝️   10   Caleb and Kelsey   

                                        track_name  duration_ms  \
0       Lose Control (feat. Ciara & Fat Man Scoop)       226863   
1                                            Toxic       198800   
2                                    Crazy In Love       235933   
3                                   Rock Your Body       267266   
4                                     It Wasn't Me       227600   
...                                            ...          ...   
198000                                   Waterfall       209573   
198001                                    The Roar       220106   
198002        Lift Your Head Weary Sinner (Chains)       224666   
198003                                We Fall Down       280960   
198004      10,000 Reasons / What a Beautiful Name       178189   

                                            album_name  
0                                         The Cookbook  
1                                          In The Zone  
2         Dangerously In Love (Alben für die Ewigkeit)  
3                                            Justified  
4                                             Hot Shot  
...                                                ...  
198000                                    Love Ran Red  
198001                                    Love Ran Red  
198002                                    Neon Steeple  
198003  How Great Is Our God: The Essential Collection  
198004          10,000 Reasons / What a Beautiful Name  

[198005 rows x 7 columns]

In [120]:
spotify["artist_w_track"] = spotify["artist_name"] + " (" + spotify["track_name"] + ")" # create new variable by combining 'artist_name' and 'track_name'
spotify

pid playlist_name  pos        artist_name  \
0            0    Throwbacks    0      Missy Elliott   
1            0    Throwbacks    1     Britney Spears   
2            0    Throwbacks    2            Beyoncé   
3            0    Throwbacks    3  Justin Timberlake   
4            0    Throwbacks    4             Shaggy   
...        ...           ...  ...                ...   
198000  999998            ✝️    6       Chris Tomlin   
198001  999998            ✝️    7       Chris Tomlin   
198002  999998            ✝️    8            Crowder   
198003  999998            ✝️    9       Chris Tomlin   
198004  999998            ✝️   10   Caleb and Kelsey   

                                        track_name  duration_ms  \
0       Lose Control (feat. Ciara & Fat Man Scoop)       226863   
1                                            Toxic       198800   
2                                    Crazy In Love       235933   
3                                   Rock Your Body       267266   
4                                     It Wasn't Me       227600   
...                                            ...          ...   
198000                                   Waterfall       209573   
198001                                    The Roar       220106   
198002        Lift Your Head Weary Sinner (Chains)       224666   
198003                                We Fall Down       280960   
198004      10,000 Reasons / What a Beautiful Name       178189   

                                            album_name  \
0                                         The Cookbook   
1                                          In The Zone   
2         Dangerously In Love (Alben für die Ewigkeit)   
3                                            Justified   
4                                             Hot Shot   
...                                                ...   
198000                                    Love Ran Red   
198001                                    Love Ran Red   
198002                                    Neon Steeple   
198003  How Great Is Our God: The Essential Collection   
198004          10,000 Reasons / What a Beautiful Name   

                                           artist_w_track  
0       Missy Elliott (Lose Control (feat. Ciara & Fat...  
1                                  Britney Spears (Toxic)  
2                                 Beyoncé (Crazy In Love)  
3                      Justin Timberlake (Rock Your Body)  
4                                   Shaggy (It Wasn't Me)  
...                                                   ...  
198000                           Chris Tomlin (Waterfall)  
198001                            Chris Tomlin (The Roar)  
198002     Crowder (Lift Your Head Weary Sinner (Chains))  
198003                        Chris Tomlin (We Fall Down)  
198004  Caleb and Kelsey (10,000 Reasons / What a Beau...  

[198005 rows x 8 columns]

In [121]:
spotify.describe( include="all") # descriptive statistics

pid playlist_name            pos artist_name track_name  \
count   198005.000000        198005  198005.000000      198005     198005   
unique            NaN          2184            NaN       18866      64115   
top               NaN       Country            NaN       Drake     Closer   
freq              NaN          3530            NaN        2715        205   
mean    323435.502775           NaN      54.391707         NaN        NaN   
std     466890.211778           NaN      48.352965         NaN        NaN   
min          0.000000           NaN       0.000000         NaN        NaN   
25%        738.000000           NaN      17.000000         NaN        NaN   
50%       1476.000000           NaN      40.000000         NaN        NaN   
75%     999210.000000           NaN      79.000000         NaN        NaN   
max     999998.000000           NaN     248.000000         NaN        NaN   

         duration_ms album_name     artist_w_track  
count   1.980050e+05     198005             198005  
unique           NaN      35497              76855  
top              NaN      Views  Drake (One Dance)  
freq             NaN        659                143  
mean    2.347408e+05        NaN                NaN  
std     1.329186e+05        NaN                NaN  
min     0.000000e+00        NaN                NaN  
25%     1.980000e+05        NaN                NaN  
50%     2.246930e+05        NaN                NaN  
75%     2.585330e+05        NaN                NaN  
max     2.074458e+07        NaN                NaN

In [122]:
spotify['artist_w_track'].value_counts() # count 'artist_w_track' to identify most popular artists & songs

Drake (One Dance)                                             143
Kendrick Lamar (HUMBLE.)                                      142
The Chainsmokers (Closer)                                     129
DRAM (Broccoli (feat. Lil Yachty))                            127
Post Malone (Congratulations)                                 119
                                                             ... 
Ellie Goulding (Anything Could Happen - Betablock3r Remix)      1
Charlie Heat (Dirty Taurus)                                     1
Russ (For the Stunt)                                            1
Tee Grizzley (Secrets)                                          1
Caleb and Kelsey (10,000 Reasons / What a Beautiful Name)       1
Name: artist_w_track, Length: 76855, dtype: int64

# Question 6
- Create a DataFrame that only contains observations from playlists featuring the song “One Dance” by Drake.
- Next, identify the song with the highest popularity after Drake’s “One Dance” in the newly created DataFrame.

*Answer*

**The song with the highest popularity after Drake's "One Dance" is "Closer" at 205.**

In [142]:
spotify_2 = spotify[['playlist_name','artist_name', 'track_name', 'artist_w_track']] # create a new DataFrame
spotify_2

playlist_name        artist_name  \
0         Throwbacks      Missy Elliott   
1         Throwbacks     Britney Spears   
2         Throwbacks            Beyoncé   
3         Throwbacks  Justin Timberlake   
4         Throwbacks             Shaggy   
...              ...                ...   
198000            ✝️       Chris Tomlin   
198001            ✝️       Chris Tomlin   
198002            ✝️            Crowder   
198003            ✝️       Chris Tomlin   
198004            ✝️   Caleb and Kelsey   

                                        track_name  \
0       Lose Control (feat. Ciara & Fat Man Scoop)   
1                                            Toxic   
2                                    Crazy In Love   
3                                   Rock Your Body   
4                                     It Wasn't Me   
...                                            ...   
198000                                   Waterfall   
198001                                    The Roar   
198002        Lift Your Head Weary Sinner (Chains)   
198003                                We Fall Down   
198004      10,000 Reasons / What a Beautiful Name   

                                           artist_w_track  
0       Missy Elliott (Lose Control (feat. Ciara & Fat...  
1                                  Britney Spears (Toxic)  
2                                 Beyoncé (Crazy In Love)  
3                      Justin Timberlake (Rock Your Body)  
4                                   Shaggy (It Wasn't Me)  
...                                                   ...  
198000                           Chris Tomlin (Waterfall)  
198001                            Chris Tomlin (The Roar)  
198002     Crowder (Lift Your Head Weary Sinner (Chains))  
198003                        Chris Tomlin (We Fall Down)  
198004  Caleb and Kelsey (10,000 Reasons / What a Beau...  

[198005 rows x 4 columns]

In [143]:
spotify_2.set_index("artist_w_track", inplace = True) # set a new index

In [144]:
spotify_2.loc[ "Drake (One Dance)" ]

playlist_name artist_name track_name
artist_w_track                                         
Drake (One Dance)           boat       Drake  One Dance
Drake (One Dance)   party people       Drake  One Dance
Drake (One Dance)  w o r k o u t       Drake  One Dance
Drake (One Dance)        English       Drake  One Dance
Drake (One Dance)   Summer Chill       Drake  One Dance
...                          ...         ...        ...
Drake (One Dance)           mixx       Drake  One Dance
Drake (One Dance)       (Guilty)       Drake  One Dance
Drake (One Dance)       Marathon       Drake  One Dance
Drake (One Dance)      old stuff       Drake  One Dance
Drake (One Dance)          Drake       Drake  One Dance

[143 rows x 3 columns]

In [145]:
spotify_2['track_name'].value_counts() # count most popular song after Drake's 'One Dance'

Closer                                                                            205
One Dance                                                                         144
HUMBLE.                                                                           142
Home                                                                              136
Roses                                                                             132
                                                                                 ... 
Make It Home                                                                        1
So Fast, So Maybe                                                                   1
The Addams Family: Main Theme (from the Television Series "The Addams Family")      1
This House Is Haunted                                                               1
10,000 Reasons / What a Beautiful Name                                              1
Name: track_name, Length: 64115, dtype: int64